### PR TITLE
controller/kubelet-config: Change validation of KubeletConfiguration to compare type-cast fields instead of reflections.

### PR DESCRIPTION
**Pull request by LorbusChris**
**- What I did**
Change validation of KubeletConfiguration to compare to type-cast fields instead of reflections.

**- How to verify it**
unit tests and CI

**- Description for the changelog**
Change validation of KubeletConfiguration to compare to type-cast fields instead of reflections.

### CODE REVIEW

File: pkg/controller/kubelet-config/helpers.go [hunk 1]
```diff
@@ -74,47 +72,41 @@ func validateUserKubeletConfig(cfg *mcfgv1.KubeletConfig) error {
...
+		return fmt.Errorf("KubeletConfiguration: ClusterDomain is not allowed to be set")
+	}
+
+	// FeatureGates map[string]bool `json:"featureGates,omitempty"`
+	if len(kcDecoded.FeatureGates) > 0 {
+		return fmt.Errorf("KubeletConfiguration: FeatureGates is not allowed to be set")
+	}
+
+	// RuntimeRequestTimeout metav1.Duration `json:"runtimeRequestTimeout,omitempty"`
+	if kcDecoded.RuntimeRequestTimeout.String() != "&Duration{Duration:0s,}" {
```
Comment (LorbusChris) [inline pkg/controller/kubelet-config/helpers.go]: I assume there is a prettier way for this

Comment (runcom) [reply pkg/controller/kubelet-config/helpers.go]: `.IsZero()`

Comment (runcom) [reply pkg/controller/kubelet-config/helpers.go]: ok just compare to zero and it works

File: pkg/controller/kubelet-config/helpers.go [hunk 2]
```diff
@@ -74,47 +72,41 @@ func validateUserKubeletConfig(cfg *mcfgv1.KubeletConfig) error {
...
+	if kcDecoded.ClusterDomain != "" {
+		return fmt.Errorf("KubeletConfiguration: ClusterDomain is not allowed to be set")
+	}
+
+	// FeatureGates map[string]bool `json:"featureGates,omitempty"`
+	if len(kcDecoded.FeatureGates) > 0 {
+		return fmt.Errorf("KubeletConfiguration: FeatureGates is not allowed to be set")
+	}
+
+	// RuntimeRequestTimeout metav1.Duration `json:"runtimeRequestTimeout,omitempty"`
```
Comment (runcom) [inline pkg/controller/kubelet-config/helpers.go]: can you drop all these comments? 👼

Comment (LorbusChris) [reply pkg/controller/kubelet-config/helpers.go]: sure. thought it would be nice to know what we are operating on here without looking it up. removing

**Comment by LorbusChris** [comment]
let's put a 
/hold
on this until both @sttts and @rphillips were able to get a look :smiley:

File: pkg/controller/kubelet-config/helpers.go [hunk 3]
```diff
@@ -74,47 +72,35 @@ func validateUserKubeletConfig(cfg *mcfgv1.KubeletConfig) error {
...
+		return fmt.Errorf("KubeletConfiguration: ClusterDomain is not allowed to be set, but contains: %s", kcDecoded.ClusterDomain)
+	}
+	if len(kcDecoded.FeatureGates) > 0 {
+		return fmt.Errorf("KubeletConfiguration: FeatureGates is not allowed to be set, but contains: %v", kcDecoded.FeatureGates)
+	}
+	if kcDecoded.RuntimeRequestTimeout.Duration != 0 {
+		return fmt.Errorf("KubeletConfiguration: RuntimeRequestTimeout is not allowed to be set, but contains: %s", kcDecoded.RuntimeRequestTimeout.Duration)
+	}
+	if kcDecoded.StaticPodPath != "" {
+		return fmt.Errorf("KubeletConfiguration: StaticPodPath is not allowed to be set, but contains: %s", kcDecoded.StaticPodPath)
```
Comment (sttts) [inline pkg/controller/kubelet-config/helpers.go]: much better.

nit: lower-case field name as in yaml/json

File: pkg/controller/kubelet-config/helpers.go [hunk 4]
```diff
@@ -74,47 +72,35 @@ func validateUserKubeletConfig(cfg *mcfgv1.KubeletConfig) error {
...
+	// If a user were to set these values, the system may become unrecoverable
+	// (ie: not recover after a reboot).
+	// Therefore, if the KubeletConfig CR instance contains a non-zero or non-empty value
+	// for one of the following fields, the MCC will not apply the CR and error out instead:
+	// CgroupDriver
+	// ClusterDNS
+	// ClusterDomain
+	// FeatureGates
+	// RuntimeRequestTimeout
+	// StaticPodPath
```
Comment (sttts) [inline pkg/controller/kubelet-config/helpers.go]: no need for this list. The same is visible below.

**Comment by sttts** [comment]
some nits. Looks good otherwise.

**Comment by rphillips** [comment]
nice! lgtm.

**Comment by LorbusChris** [comment]
/hold cancel
@runcom for a go

**Pull request merged**